sklearn/linear_model/bayes.py

```diff
@@ -217,45 +236,62 @@ def fit(self, X, y, sample_weight=None):
                     logdet_sigma_[:n_samples] += alpha_ * eigen_vals_
                     logdet_sigma_ = - np.sum(np.log(logdet_sigma_))
 
-            # Preserve the alpha and lambda values that were used to
-            # calculate the final coefficients
-            self.alpha_ = alpha_
-            self.lambda_ = lambda_
-
-            # Update alpha and lambda
             rmse_ = np.sum((y - np.dot(X, coef_)) ** 2)
-            gamma_ = (np.sum((alpha_ * eigen_vals_) /
-                      (lambda_ + alpha_ * eigen_vals_)))
+            # Compute the log marginal likelihood
+            if self.compute_score:
+                s = self._log_marginal_likelihood(n_samples, n_features,
+                                                  alpha_, lambda_,
+                                                  coef_, rmse_, logdet_sigma_)
+                self.scores_.append(s)
+
+            # Update alpha and lambda according to (MacKay, 1992)
+            gamma_ = np.sum((alpha_ * eigen_vals_) /
+                            (lambda_ + alpha_ * eigen_vals_))
             lambda_ = ((gamma_ + 2 * lambda_1) /
                        (np.sum(coef_ ** 2) + 2 * lambda_2))
             alpha_ = ((n_samples - gamma_ + 2 * alpha_1) /
                       (rmse_ + 2 * alpha_2))
 
-            # Compute the objective function
-            if self.compute_score:
-                s = lambda_1 * log(lambda_) - lambda_2 * lambda_
-                s += alpha_1 * log(alpha_) - alpha_2 * alpha_
-                s += 0.5 * (n_features * log(lambda_) +
-                            n_samples * log(alpha_) -
-                            alpha_ * rmse_ -
-                            (lambda_ * np.sum(coef_ ** 2)) -
-                            logdet_sigma_ -
-                            n_samples * log(2 * np.pi))
-                self.scores_.append(s)
-
             # Check for convergence
             if iter_ != 0 and np.sum(np.abs(coef_old_ - coef_)) < self.tol:
                 if verbose:
                     print("Convergence after ", str(iter_), " iterations")
                 break
             coef_old_ = np.copy(coef_)
 
+        self.n_iter_ = iter_ + 1
+
+        # return regularization parameters
+        self.alpha_ = alpha_
+        self.lambda_ = lambda_
+        # and corresponding posterior mean and posterior covariance
```

Comment:
Here you recompute the posterior mean and posterior covariance, exactly as in the for-loop.
Could you refactor the two computations into a function, to avoid duplication?
And don't you need to update `logdet_sigma_` also?

Reply:
> And don't you need to update `logdet_sigma_` also?

Indeed and and `rmse_` as well. Thanks!